### ISSUE TITLE
Slow complex-valued einsum on GPU

### ISSUE THREAD

**Issue by XDongiang** [google/jax#1763]
it is my fault
```python
import tools as tl 
from jax import device_put
import numpy as onp
import jax.numpy as np
from jax import vmap
from functools import partial
import time
from jax import jit
from jax import grad

def invm_plus(Pb,Pc):
    Pbc = Pb + Pc
    _Pbc = Pbc * np.array([-1,-1,-1,1])
    return np.sum(Pbc * _Pbc,axis=1)

def invm(Pbc):
    _Pbc = Pbc * np.array([-1,-1,-1,1])
    return np.sum(Pbc * _Pbc,axis=1)

def BW(m_,w_,Sbc):
    i = complex(0,1)
    gamma=np.sqrt(m_*m_*(m_*m_+w_*w_))
    k = np.sqrt(2*np.sqrt(2)*m_*np.abs(w_)*gamma/np.pi/np.sqrt(m_*m_+gamma))
    return k/(m_*m_ - Sbc - i*m_*w_)

def phase(theta, rho):
    i = complex(0,1)
    return rho * np.exp(theta*i)

def _abs(bw_):
    conjbw = np.conj(bw_)
    return np.real(bw_*conjbw

Kp = onp.random.sample(80000*4).reshape(80000,4)
Km = onp.random.sample(80000*4).reshape(80000,4)
Pip = onp.random.sample(80000*4).reshape(80000,4)
Pim = onp.random.sample(80000*4).reshape(80000,4)
phif001 = onp.random.sample(80000*2).reshape(80000,2)
phif021 = onp.random.sample(80000*2).reshape(80000,2)

phif0 = np.asarray([phif001,phif021])
phi = invm_plus(Kp,Km)
f0 = invm_plus(Pip,Pim)

phim = np.array([2.,1.,1.,1.,2.,1.,1.,1.])
phiw = np.array([1.,2.,1.,1.,2.,1.,1.,1.])
f0m = np.array([1.,1.,1.,3.,1.,1.,1.,1.])
f0w = np.array([1.,1.,1.,1.,1.,1.,1.,1.])
const = np.array([[2.,1.,1.,1.,1.,1.,1.,1.],[1.,1.,1.,1.,1.,1.,1.,1.]])
rho = np.array([1.,1.,2.,1.,1.,1.,1.,1.])
theta = np.array([1.,1.,1.,1.,3.,1.,3.,1.])


def BW_f0(phim,phiw,f0m,f0w,phi,f0):
    return vmap(partial(BW,Sbc=phi))(phim,phiw) * vmap(partial(BW,Sbc=f0))(f0m,f0w)

def phase_f0(theta_,rho_):
    result = vmap(phase)(theta_,rho_)
    return result

def test_pw(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0):
    ph = phase_f0(theta,rho)
    bw = BW_f0(phim,phiw,f0m,f0w,phi,f0)
    const_phase = np.einsum('ij,j->ij',const,ph)
    _phif0 = np.einsum('ijk,il->ljk',phif0,const_phase)
    _phif0 = np.einsum('ijk,ij->jk',_phif0,bw)
    _phif0 = np.real(np.sum(_abs(_phif0),axis=1))
    return -np.sum(np.log(_phif0))

test_pw_jit = jit(test_pw)

print(test_pw_jit(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0))

m = (0,1,2,3,4,5,6)
grad_test_pw = grad(test_pw_jit,argnums=m)
s = time.time()
print(grad_test_pw(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0))
e = time.time()
print("time : ", e - s)

def test_pw1(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0):
    temp = np.zeros(80000)
    for i in range(8):
        bw = BW(phim[i],phiw[i],phi) * BW(f0m[i],f0w[i],f0)
        _phif001 = phif001.T * bw *const[0,i]
        _phif021 = phif021.T * bw *const[1,i]
        _phif0 = (_phif001 + _phif021) * phase(theta[i],rho[i])
        temp = temp + _phif0
    data = np.real(np.sum(_abs(temp),axis=0))
    return -np.sum(np.log(data))

test_pw1_jit = jit(test_pw1)   
grad_test_pw1 = grad(test_pw1_jit,argnums=m)

print(test_pw1_jit(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0))
s = time.time()
print(grad_test_pw1(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0))
e = time.time()
print("time : ", e - s)
```
it is my real code , i test it is the same output value

**Comment by XDongiang** [comment]
my device is GTX1060 6G and i7 8750H

**Comment by hawkinsp** [comment]
Two quick comments:
a) JAX uses asynchronous dispatch, which means you need to be careful when writing timing microbenchmarks. Please read: https://jax.readthedocs.io/en/latest/async_dispatch.html . I suspect you are not timing the computation itself at all here, instead only the time to dispatch it.

b) the two computations don't appear to compute the same thing (i.e., they return different values). So it seems meaningless to compare their timings? Can you make the two computations identical?

**Comment by XDongiang** [comment]
> Two quick comments:
> a) JAX uses asynchronous dispatch, which means you need to be careful when writing timing microbenchmarks. Please read: https://jax.readthedocs.io/en/latest/async_dispatch.html . I suspect you are not timing the computation itself at all here, instead only the time to dispatch it.
> 
> b) the two computations don't appear to compute the same thing (i.e., they return different values). So it seems meaningless to compare their timings? Can you make the two computations identical?

thank your replay , but when i use onp.asarray() timing the program , the time is also same

**Comment by hawkinsp** [comment]
There's two more suggestions I'd make about the benchmark here:
*  the first time you invoke a `jit`-wrapped computation, JAX compiles the computation using XLA. Compilation is cached. To avoid measuring compilation time, you should perform an untimed warmup of each computation before a timed run. This makes a very significant difference to the timings your script prints!
* it's probably faster to stick a `jit(...)` decorator on the outside of your `grad(...)` rather than the other way around.

That said, once done, I also find that the `einsum` version is surprisingly slow. I think what is happening here is that XLA is missing a strength reduction optimization for batched matrix-vector multiplications for complex types, and is currently falling back to a naive slow implementation. I raised an issue with the XLA team about this.

Thanks for the report!

**Comment by XDongiang** [comment]
You say it is surprisingly slow , so i want to konw what method is fast than eisum when do the same ting

**Comment by XDongiang** [comment]
I test use real value to calculate 
```python
def BW(m_,w_,Sbc):
#     i = complex(0,1)
    gamma=np.sqrt(m_*m_*(m_*m_+w_*w_))
    k = np.sqrt(2*np.sqrt(2)*m_*np.abs(w_)*gamma/np.pi/np.sqrt(m_*m_+gamma))
    return k/(m_*m_ - Sbc - m_*w_)
def test_pw1(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0):
    temp = np.zeros(800000)
    for i in range(8):
        bw = BW(phim[i],phiw[i],phi) * BW(f0m[i],f0w[i],f0)
        _phif001 = phif001.T * bw *const[0,i]
        _phif021 = phif021.T * bw *const[1,i]
        _phif0 = (_phif001 + _phif021) * phase(theta[i],rho[i])
        temp = temp + _phif0
    data = np.real(np.sum(_abs(temp),axis=0))
    return -np.sum(np.log(data))
test_pw1_jit = jit(test_pw1)                   
grad_test_pw = jit(grad(test_pw,argnums=m))
%timeit onp.asarray(grad_test_pw(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0))
406 ms ± 1.34 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
The difference of two methods is not too big.

**Comment by XDongiang** [comment]
> There's two more suggestions I'd make about the benchmark here:
> 
> * the first time you invoke a `jit`-wrapped computation, JAX compiles the computation using XLA. Compilation is cached. To avoid measuring compilation time, you should perform an untimed warmup of each computation before a timed run. This makes a very significant difference to the timings your script prints!
> * it's probably faster to stick a `jit(...)` decorator on the outside of your `grad(...)` rather than the other way around.
> 
> That said, once done, I also find that the `einsum` version is surprisingly slow. I think what is happening here is that XLA is missing a strength reduction optimization for batched matrix-vector multiplications for complex types, and is currently falling back to a naive slow implementation. I raised an issue with the XLA team about this.
> 
> Thanks for the report!

hi , it is a very long time.  Slow complex-valued einsum on GPU , it still a problem ?

**Comment by hawkinsp** [comment]
Taking a look again, it looks to me like this issue is resolved at head.

My version of the benchmark looks like this:
```
from jax import device_put
import numpy as onp
import jax.numpy as np
from jax import vmap
from functools import partial
import time
from jax import jit
from jax import grad

def invm_plus(Pb,Pc):
    Pbc = Pb + Pc
    _Pbc = Pbc * np.array([-1,-1,-1,1])
    return np.sum(Pbc * _Pbc,axis=1)

def invm(Pbc):
    _Pbc = Pbc * np.array([-1,-1,-1,1])
    return np.sum(Pbc * _Pbc,axis=1)

def BW(m_,w_,Sbc):
    i = complex(0,1)
    gamma=np.sqrt(m_*m_*(m_*m_+w_*w_))
    k = np.sqrt(2*np.sqrt(2)*m_*np.abs(w_)*gamma/np.pi/np.sqrt(m_*m_+gamma))
    return k/(m_*m_ - Sbc - i*m_*w_)

def phase(theta, rho):
    i = complex(0,1)
    return rho * np.exp(theta*i)

def _abs(bw_):
    conjbw = np.conj(bw_)
    return np.real(bw_*conjbw)

Kp = onp.random.sample(80000*4).reshape(80000,4)
Km = onp.random.sample(80000*4).reshape(80000,4)
Pip = onp.random.sample(80000*4).reshape(80000,4)
Pim = onp.random.sample(80000*4).reshape(80000,4)
phif001 = onp.random.sample(80000*2).reshape(80000,2)
phif021 = onp.random.sample(80000*2).reshape(80000,2)

phif0 = np.asarray([phif001,phif021])
phi = invm_plus(Kp,Km)
f0 = invm_plus(Pip,Pim)

phim = np.array([2.,1.,1.,1.,2.,1.,1.,1.])
phiw = np.array([1.,2.,1.,1.,2.,1.,1.,1.])
f0m = np.array([1.,1.,1.,3.,1.,1.,1.,1.])
f0w = np.array([1.,1.,1.,1.,1.,1.,1.,1.])
const = np.array([[2.,1.,1.,1.,1.,1.,1.,1.],[1.,1.,1.,1.,1.,1.,1.,1.]])
rho = np.array([1.,1.,2.,1.,1.,1.,1.,1.])
theta = np.array([1.,1.,1.,1.,3.,1.,3.,1.])


def BW_f0(phim,phiw,f0m,f0w,phi,f0):
    return vmap(partial(BW,Sbc=phi))(phim,phiw) * vmap(partial(BW,Sbc=f0))(f0m,f0w)

def phase_f0(theta_,rho_):
    result = vmap(phase)(theta_,rho_)
    return result

def test_pw(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0):
    ph = phase_f0(theta,rho)
    bw = BW_f0(phim,phiw,f0m,f0w,phi,f0)
    const_phase = np.einsum('ij,j->ij',const,ph)
    _phif0 = np.einsum('ijk,il->ljk',phif0,const_phase)
    _phif0 = np.einsum('ijk,ij->jk',_phif0,bw)
    _phif0 = np.real(np.sum(_abs(_phif0),axis=1))
    return -np.sum(np.log(_phif0))

test_pw_jit = jit(test_pw)

print(test_pw_jit(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0))

m = (0,1,2,3,4,5,6)
grad_test_pw = jit(grad(test_pw_jit,argnums=m))
_ = grad_test_pw(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0)[0].block_until_ready()
s = time.time()
print(grad_test_pw(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0))
e = time.time()
print("time : ", e - s)

def test_pw1(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0):
    temp = np.zeros(80000)
    for i in range(8):
        bw = BW(phim[i],phiw[i],phi) * BW(f0m[i],f0w[i],f0)
        _phif001 = phif001.T * bw *const[0,i]
        _phif021 = phif021.T * bw *const[1,i]
        _phif0 = (_phif001 + _phif021) * phase(theta[i],rho[i])
        temp = temp + _phif0
    data = np.real(np.sum(_abs(temp),axis=0))
    return -np.sum(np.log(data))

grad_test_pw1 = jit(grad(test_pw1,argnums=m))

#print(test_pw1_jit(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0))
_ = grad_test_pw1(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0)[0].block_until_ready()
s = time.time()
print(grad_test_pw1(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0))
e = time.time()
print("time : ", e - s)
```

Both versions seem roughly equal in performance, around 4-5ms on a P100.

**Comment by XDongiang** [comment]
> Taking a look again, it looks to me like this issue is resolved at head.
> 
> My version of the benchmark looks like this:
> 
> ```
> from jax import device_put
> import numpy as onp
> import jax.numpy as np
> from jax import vmap
> from functools import partial
> import time
> from jax import jit
> from jax import grad
> 
> def invm_plus(Pb,Pc):
>     Pbc = Pb + Pc
>     _Pbc = Pbc * np.array([-1,-1,-1,1])
>     return np.sum(Pbc * _Pbc,axis=1)
> 
> def invm(Pbc):
>     _Pbc = Pbc * np.array([-1,-1,-1,1])
>     return np.sum(Pbc * _Pbc,axis=1)
> 
> def BW(m_,w_,Sbc):
>     i = complex(0,1)
>     gamma=np.sqrt(m_*m_*(m_*m_+w_*w_))
>     k = np.sqrt(2*np.sqrt(2)*m_*np.abs(w_)*gamma/np.pi/np.sqrt(m_*m_+gamma))
>     return k/(m_*m_ - Sbc - i*m_*w_)
> 
> def phase(theta, rho):
>     i = complex(0,1)
>     return rho * np.exp(theta*i)
> 
> def _abs(bw_):
>     conjbw = np.conj(bw_)
>     return np.real(bw_*conjbw)
> 
> Kp = onp.random.sample(80000*4).reshape(80000,4)
> Km = onp.random.sample(80000*4).reshape(80000,4)
> Pip = onp.random.sample(80000*4).reshape(80000,4)
> Pim = onp.random.sample(80000*4).reshape(80000,4)
> phif001 = onp.random.sample(80000*2).reshape(80000,2)
> phif021 = onp.random.sample(80000*2).reshape(80000,2)
> 
> phif0 = np.asarray([phif001,phif021])
> phi = invm_plus(Kp,Km)
> f0 = invm_plus(Pip,Pim)
> 
> phim = np.array([2.,1.,1.,1.,2.,1.,1.,1.])
> phiw = np.array([1.,2.,1.,1.,2.,1.,1.,1.])
> f0m = np.array([1.,1.,1.,3.,1.,1.,1.,1.])
> f0w = np.array([1.,1.,1.,1.,1.,1.,1.,1.])
> const = np.array([[2.,1.,1.,1.,1.,1.,1.,1.],[1.,1.,1.,1.,1.,1.,1.,1.]])
> rho = np.array([1.,1.,2.,1.,1.,1.,1.,1.])
> theta = np.array([1.,1.,1.,1.,3.,1.,3.,1.])
> 
> 
> def BW_f0(phim,phiw,f0m,f0w,phi,f0):
>     return vmap(partial(BW,Sbc=phi))(phim,phiw) * vmap(partial(BW,Sbc=f0))(f0m,f0w)
> 
> def phase_f0(theta_,rho_):
>     result = vmap(phase)(theta_,rho_)
>     return result
> 
> def test_pw(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0):
>     ph = phase_f0(theta,rho)
>     bw = BW_f0(phim,phiw,f0m,f0w,phi,f0)
>     const_phase = np.einsum('ij,j->ij',const,ph)
>     _phif0 = np.einsum('ijk,il->ljk',phif0,const_phase)
>     _phif0 = np.einsum('ijk,ij->jk',_phif0,bw)
>     _phif0 = np.real(np.sum(_abs(_phif0),axis=1))
>     return -np.sum(np.log(_phif0))
> 
> test_pw_jit = jit(test_pw)
> 
> print(test_pw_jit(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0))
> 
> m = (0,1,2,3,4,5,6)
> grad_test_pw = jit(grad(test_pw_jit,argnums=m))
> _ = grad_test_pw(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0)[0].block_until_ready()
> s = time.time()
> print(grad_test_pw(phim,phiw,f0m,f0w,const,rho,theta,phif0,phi,f0))
> e = time.time()
> print("time : ", e - s)
> 
> def test_pw1(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0):
>     temp = np.zeros(80000)
>     for i in range(8):
>         bw = BW(phim[i],phiw[i],phi) * BW(f0m[i],f0w[i],f0)
>         _phif001 = phif001.T * bw *const[0,i]
>         _phif021 = phif021.T * bw *const[1,i]
>         _phif0 = (_phif001 + _phif021) * phase(theta[i],rho[i])
>         temp = temp + _phif0
>     data = np.real(np.sum(_abs(temp),axis=0))
>     return -np.sum(np.log(data))
> 
> grad_test_pw1 = jit(grad(test_pw1,argnums=m))
> 
> #print(test_pw1_jit(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0))
> _ = grad_test_pw1(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0)[0].block_until_ready()
> s = time.time()
> print(grad_test_pw1(phim,phiw,f0m,f0w,theta,rho,const,phif001,phif021,phi,f0))
> e = time.time()
> print("time : ", e - s)
> ```
> 
> Both versions seem roughly equal in performance, around 4-5ms on a P100.

Thank you for you reply. 
Please tell me the JAX version you are using.
And cuda and cudnn use which version is better?
